compiler/rustc_infer/src/infer/outlives/verify.rs

```diff
@@ -119,66 +133,49 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
     /// the clause from the environment only applies if `'0 = 'a`,
     /// which we don't know yet. But we would still include `'b` in
     /// this list.
-    pub fn projection_approx_declared_bounds_from_env(
+    pub fn approx_declared_bounds_from_env(
         &self,
-        projection_ty: ty::ProjectionTy<'tcx>,
+        generic: GenericKind<'tcx>,
     ) -> Vec<ty::OutlivesPredicate<Ty<'tcx>, ty::Region<'tcx>>> {
-        let projection_ty = GenericKind::Projection(projection_ty).to_ty(self.tcx);
-        let erased_projection_ty = self.tcx.erase_regions(projection_ty);
+        let generic_ty = generic.to_ty(self.tcx);
+        let erased_ty = self.tcx.erase_regions(generic_ty);
+        // We add a fast path for the discriminant to not erase all types before comparing,
+        // but only the ones that could possibly be a match.
+        let discr = std::mem::discriminant(erased_ty.kind());
         self.declared_generic_bounds_from_env_with_compare_fn(|ty| {
-            if let ty::Projection(..) = ty.kind() {
-                let erased_ty = self.tcx.erase_regions(ty);
-                erased_ty == erased_projection_ty
-            } else {
-                false
-            }
+            std::mem::discriminant(ty.kind()) == discr && self.tcx.erase_regions(ty) == erased_ty
         })
     }
 
-    /// Searches the where-clauses in scope for regions that
-    /// `projection_ty` is known to outlive. Currently requires an
-    /// exact match.
-    pub fn projection_declared_bounds_from_trait(
-        &self,
-        projection_ty: ty::ProjectionTy<'tcx>,
-    ) -> impl Iterator<Item = ty::Region<'tcx>> + 'cx + Captures<'tcx> {
-        self.declared_projection_bounds_from_trait(projection_ty)
-    }
-
-    pub fn projection_bound(
+    #[instrument(level = "debug", skip(self, visited))]
+    fn env_bounds(
```

Comment:
`env_bounds` sounds misleading to me, because it applies verify bounds from TraitRef and for individual substs as well. `projection_opaque_bounds` might be more suitable.